### PR TITLE
divu instruction

### PR DESCRIPTION
**Pull request by lvella**

### CODE REVIEW

File: riscv/src/compiler.rs [hunk 1]
```diff
@@ -518,6 +518,17 @@ pil{
     { Y_b6 } in { bytes };
 }
 
+pil { col witness remainder; }
+
+// implements X = Y / Z, stores remainder in `remainder`.
```
Comment (chriseth) [inline riscv/src/compiler.rs:523]: Oh we should also handle division by zero properly.

Comment (lvella) [reply riscv/src/compiler.rs:523]: What is division by zero in riscv?

Comment (lvella) [reply riscv/src/compiler.rs:523]: In our definition of division, it seems impossible to have a result if the divisor is 0. May not be the riscv mandated result, but it kinda covers it already.

**Comment by leonardoalt** [comment]
@lvella what's missing for a basic version of this PR to be merged?

**Comment by lvella** [comment]
> @lvella what's missing for a basic version of this PR to be merged?

Me to understand how to make a `RangeConstraint` in the witgen, and use to identify division operations.

File: riscv/src/compiler.rs (hunk 1 above)
Comment (lvella) [reply riscv/src/compiler.rs:545]: BTW, in riscv, division by zero sets all bits to 1.

File: riscv/build.rs [hunk 2]
```diff
@@ -29,6 +29,7 @@ fn build_instruction_tests() {
             .to_string()
             .strip_suffix(".S")
         {
+            println!("cargo:rerun-if-changed={file_name}.S");
```
Comment (chriseth) [inline riscv/build.rs:32]: Did you mean to use `write!` instead of `println!`?

File: riscv/src/compiler.rs [hunk 3]
```diff
@@ -544,11 +544,48 @@ fn preamble() -> String {
...
+    }
+
+    // implements Z = Y / X, stores remainder in `remainder`.
+    instr divu Y, X -> Z {
+        // Y is the known dividend
+        // X is the known divisor
+        // Z is the unknown quotient
+        // main division algorithm;
+        // if X is zero, remainder is set to dividend, as per RISC-V specification:
+        X * Z + remainder - Y = 0,
```
Comment (chriseth) [inline riscv/src/compiler.rs:576]:
```suggestion
        X * Z + remainder = Y;
```?

File: riscv/build.rs (hunk 2 above)
Comment (lvella) [reply riscv/build.rs:32]: No, this is a command to cargo rebuild if the assembly file changes.

File: riscv/src/compiler.rs [hunk 4]
```diff
@@ -544,11 +544,48 @@ fn preamble() -> String {
...
+
+        // remainder > 0:
```
Comment (chriseth) [inline riscv/src/compiler.rs:578]:
```suggestion
        // remainder >= 0:
```

File: riscv/src/compiler.rs (hunk 3 above)
Comment (lvella) [reply riscv/src/compiler.rs:576]: How do you do these suggest changes?

Comment (chriseth) [reply riscv/src/compiler.rs:576]: The +- icon

Comment (chriseth) [reply riscv/src/compiler.rs:576]: If you click and drag over multiple lines, you can also suggest to change multiple lines at once.

**Comment by chriseth** [comment]
Looks good! Does it work?

**Comment by lvella** [comment]
> Looks good! Does it work? :)

According to the divu instruction test, yes.

**Comment by leonardoalt** [comment]
squash & merge? @chriseth

**Review by chriseth — APPROVED** [review]

**Pull request merged**